elastalert/elastalert.py

```diff
@@ -784,7 +784,10 @@ def alert(self, matches, rule, alert_time=None):
             try:
                 alert.alert(matches)
             except EAException as e:
-                self.handle_error('Error while running alert %s: %s' % (alert.get_info()['type'], e), {'rule': rule['name']})
+                self.handle_error('ElastAlert Error while running alert %s: %s' % (alert.get_info()['type'], e), {'rule': rule['name']})
+                alert_exception = str(e)
+            except Exception as e:
+                self.handle_error('Unexpected Error while running alert %s: %s' % (alert.get_info()['type'], e), {'rule': rule['name']})
```

Comment:
You'll want to be cautious here. The class `Exception` includes `SyntaxError` and `RuntimeException`, which can make crafting new rules and identifying their failure scenarios problematic. This is a form of [Pokemon Exception Handling](http://c2.com/cgi/wiki?PokemonExceptionHandling) that may have unexpected ramifications in developing and testing alerts. This also obliterates the traceback for the error, further confounding development.

In practice, this is exactly why we created EAException: it allowed exception handling to be pushed down into the rules themselves  instead of relying on the scheduler.

Errors that crash ElastAlert itself should be as loud as possible, but I agree that they should not crash the application. An alternative to this approach would be to provide a thread pool for rules, reporting and reclaiming any crashed threads along the way. This can be done with a thread headcount of 1 to start with until single-threaded assumptions are resolved. Threading is something we want to do anyway, which further strengthens this recommendation.

In short: let's not make it easy for rules authors to write flaky rules that rely on the scheduler to handle. Doing so will incentivize low alert quality, which is counterproductive to users of this application.